Add locking and bypassing of pre-minified files to optimize the asset…

EreMaijala opened the pull request:
… pipeline cache file creation.

We have a couple of pretty large pre-minified js files (e.g. video.js and a couple of media source extensions), and running those through the minifier is slow (as in multiple seconds). When we get more than 30 requests/sec on a single front end server and all of them are trying to minify a file, possible the same one, at the same time, the server is quite badly affected (like 100 times the normal load for a couple of minutes). The locking introduced here makes sure the same file is only created once by the first one requesting it, and the other requests just wait for it to be done. Bypassing minification of large files makes the whole process quite a bit faster too (e.g. from 10 seconds to few hundred milliseconds).

demiankatz commented on module/VuFindTheme/src/VuFindTheme/View/Helper/ConcatTrait.php:
```diff
@@ -291,6 +341,21 @@ protected function outputInOrder($indent)
+    protected function isMinifiable($filename)
+    {
+        $basename = basename($filename);
+        return strpos($basename, '.min.js') === false
```
Would it be simpler to simply check for ".min." anywhere in the filename, or are you worried that is too generic? Another option might be to do a regular expression match on `\.min\.(js|css)`. Using a regular expression is slightly more concise and might also open the door to making the pattern configurable through an .ini setting in the future if there is a need to do so.

demiankatz commented on module/VuFindTheme/src/VuFindTheme/View/Helper/ConcatTrait.php:
```diff
@@ -228,15 +231,62 @@ protected function getConcatenatedFilePath($group)
+                            . $this->getResourceFilePath($item),
+                            ThemeInfo::RETURN_ALL_DETAILS
+                        );
+                        if ($this->isMinifiable($details['path'])) {
```
Would it make sense to refactor lines 255-273 into a separate getMinifiedData method? That way, you could extend the method in the HeadScript helper to put the JS-specific logic there, instead of having a special case in the shared logic. It might also help readability a bit, as this method has become rather long.

EreMaijala commented:
I believe I've now addressed the review comments. I also cleaned up the way the file is written so that the temporary file is eliminated and moved the writing to a separate method to allow it to be customized (something we need to do anyway).
